examples/name-service/programs/name-service/src/lib.rs

```diff
@@ -0,0 +1,249 @@
+    sdk::{
+        address::derive_address,
+        compressed_account::{
+            CompressedAccount, CompressedAccountData, PackedCompressedAccountWithMerkleContext,
+        },
+        CompressedCpiContext,
+    },
+    InstructionDataInvokeCpi, NewAddressParamsPacked, OutputCompressedAccountWithPackedContext,
+};
+use light_utils::hash_to_bn254_field_size_be;
+
+declare_id!("7yucc7fL3JGbyMwg4neUaenNSdySS39hbAk89Ao3t1Hz");
+
+#[program]
+pub mod name_service {
+
+    use light_sdk::utils::create_cpi_inputs_for_new_address;
+
+    use super::*;
+
+    pub fn create_record<'info>(
+        ctx: Context<'_, '_, '_, 'info, NameService<'info>>,
+        proof: CompressedProof,
+        new_address_params: NewAddressParamsPacked,
+        name: String,
+        rdata: RData,
+        cpi_context: Option<CompressedCpiContext>,
+    ) -> Result<()> {
+        let record = NameRecord {
+            owner: ctx.accounts.signer.key(),
+            name,
+            rdata,
+        };
+        let compressed_account = create_compressed_account(&ctx, &record, &new_address_params)?;
+
+        let signer_seed = b"cpi_signer".as_slice();
+        let bump = Pubkey::find_program_address(&[signer_seed], &ctx.accounts.self_program.key()).1;
+        let signer_seeds = [signer_seed, &[bump]];
+
+        let inputs = create_cpi_inputs_for_new_address(
+            proof,
+            new_address_params,
+            compressed_account,
+            &signer_seeds,
+            cpi_context,
+        );
+
+        verify(ctx, &inputs, &[&signer_seeds])?;
+
+        Ok(())
+    }
+
+    pub fn update_record<'info>(
+        ctx: Context<'_, '_, '_, 'info, NameService<'info>>,
+        compressed_account: PackedCompressedAccountWithMerkleContext,
+        proof: CompressedProof,
+        address: [u8; 32],
+        name: String,
+        rdata: RData,
+        cpi_context: Option<CompressedCpiContext>,
+    ) -> Result<()> {
+        signer_check(&ctx, &compressed_account)?;
+
+        let record = NameRecord {
+            owner: ctx.accounts.signer.key(),
+            name,
+            rdata,
+        };
+        let new_compressed_account = compressed_account_with_address(&record, address)?;
+
+        let signer_seed = b"cpi_signer".as_slice();
+        let bump = Pubkey::find_program_address(&[signer_seed], &ctx.accounts.self_program.key()).1;
+        let signer_seeds = [signer_seed, &[bump]];
+
+        let inputs = InstructionDataInvokeCpi {
+            proof: Some(proof),
+            new_address_params: vec![],
+            input_compressed_accounts_with_merkle_context: vec![compressed_account],
+            output_compressed_accounts: vec![new_compressed_account],
+            relay_fee: None,
+            compress_or_decompress_lamports: None,
+            is_compress: false,
+            signer_seeds: signer_seeds.iter().map(|seed| seed.to_vec()).collect(),
+            cpi_context,
+        };
+
+        verify(ctx, &inputs, &[&signer_seeds])?;
```

Comment:
why do we pass signer seeds here and in the inputs struct?

Reply:
I wasn't the one writing the `verify` function, but I think it was done this way because of type differences in how we and Anchor store seeds.

We store them in `InstructionDataInvokeCpi` as vectors:

```rust
pub struct InstructionDataInvokeCpi {
    [...]
    pub signer_seeds: Vec<Vec<u8>>,
    [...]
}
```

Anchor expects them as slices, (they really like declaring lifetimes). `CpiContext::new_with_signer` looks like:

```rust
pub struct CpiContext<'a, 'b, 'c, 'info, T>
where
    T: ToAccountMetas + ToAccountInfos<'info>,
{
    pub accounts: T,
    pub remaining_accounts: Vec<AccountInfo<'info>>,
    pub program: AccountInfo<'info>,
    pub signer_seeds: &'a [&'b [&'c [u8]]],
}

[...]

    #[must_use]
    pub fn new_with_signer(
        program: AccountInfo<'info>,
        accounts: T,
        signer_seeds: &'a [&'b [&'c [u8]]],
    ) -> Self {
        Self {
            accounts,
            program,
            signer_seeds,
            remaining_accounts: Vec::new(),
        }
    }
```

(funny that they use vectors for remaining accounts and slices for seeds)

What happens here is that we pass the duplicate signer seeds as slices, like Anchor expects.

I see two potential solutions here:

* Solution 1: Converting our `Vec<Vec<u8>>` to slices with `iter().collect()`.
  * Would be easy, I could just do this in this PR.
  * But I don't like that solution, since it's a wasteful loop. It's not performance critical, but still ugly and blows our code a bit.
* Solution 2: Change the `signer_seeds` field in `InstructionDataInvokeCpi` to `&'a [&'b [&'c [u8]]]` or `&'a [&'b [u8]]`.
  * I'm leaning towards this solution - it's cleaner and if we do it properly, we can get rid of all wasteful `iter().collect()` calls.
  * I would prefer to make it as a separate PR, because it affects also other examples and tests.

Reply:
wrt Solution 2, I think this didn't work for me because we need to copy the data to serialize it but great if you can make it work. The annoying thing is that this is the api from the system program so it would incur changes in many places so in case it works let's still evaluate it in a separate pr.